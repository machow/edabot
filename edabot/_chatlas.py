from __future__ import annotations
import contextvars
from chatlas import Chat 
from chatlas._display import MarkdownDisplay, IPyMarkdownDisplay

ctx_display: contextvars.ContextVar[IPyMarkdownDisplay] = contextvars.ContextVar("display")


class IPyMarkdownDisplay2(IPyMarkdownDisplay):
    _token: contextvars.Token | None = None

    def _init_display(self) -> str:
        # NOTE: now reset content
        self.content = ""

        try:
            from IPython.display import HTML, Markdown, display
        except ImportError:
            raise ImportError(
                "The IPython package is required for displaying content in a Jupyter notebook. "
                "Install it with `pip install ipython`."
            )

        # NOTE: removed the displayed, malformed HTML
        if self._css_styles:
            pass

        handle = display(Markdown(""), display_id=True)
        if handle is None:
            raise ValueError("Failed to create display handle")

        # NOTE: now assign display id
        self._ipy_display_id = handle.display_id
        return handle.display_id

    def __enter__(self):
        self._token = ctx_display.set(self)
        return super().__enter__()
    
    def __exit__(self, exc_type, exc_value, traceback):
        if self._token is not None:
            ctx_display.reset(self._token)

        
class Chat2(Chat):
    def _markdown_display(self, *args, **kwargs) -> MarkdownDisplay:
        display = super()._markdown_display(*args, **kwargs)
        if isinstance(display, IPyMarkdownDisplay):
            return IPyMarkdownDisplay2(self._echo_options)
        
        return display