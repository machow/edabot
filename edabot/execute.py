from __future__ import annotations

import traceback
from chatlas.types import (
    Content,
    ContentImageInline,
    ContentText,
    ContentToolResult,
    ContentJson,
)
from dataclasses import dataclass
from typing import Any, Union, TYPE_CHECKING
from typing_extensions import TypeAlias

if TYPE_CHECKING:
    from IPython.core.interactiveshell import InteractiveShell, ExecutionResult
    from IPython.utils.capture import RichOutput, CapturedIO

CellException: TypeAlias = "tuple[type, BaseException, Any]"


@dataclass
class CellRun:
    """The result of executing an IPython cell.

    Attributes
    ----------
    captured:
        The captured stdout, stderr, and rich outputs.
    result:
        The returned value of the cell (resulting from the last expression).
    execution_info:
        Information about execution.
    """

    captured: CapturedIO
    result: ExecutionResult
    execution_info: Union[CellException, None]

    def to_content(self) -> list[Content]:
        return list(map(self.output_to_content, self.captured.outputs))

    def to_tool_result(self) -> ContentToolResult:
        if self.execution_info is not None:
            tb = traceback.format_exception(*self.execution_info)
        else:
            tb = []
        cell_std = ContentJson(
            {
                "type": "stdout, stderr, and exceptions",
                "stdout": self.captured.stdout,
                "stderr": self.captured.stderr,
                "raw_cell": self.result.info.raw_cell,
                # TODO: might want traceback
                "error": bool(tb),
                "error_traceback": tb,
            }
        )
        cell_displays = self.to_content()
        return ContentToolResult("", [cell_std, *cell_displays])

    def to_html(self) -> str:
        tmpl = """
        <details>
          <summary>Execute Code</summary>
          <h3>Code</h3>
          <pre>{code}</pre>
          <h3>Outputs</h3>
          <pre>
          {outputs}
          </pre>

          <h3>Display</h3>
          {results}
        </details>
        """
        display = "\n\n".join(map(self.output_to_html, self.captured.outputs))

        outputs = self.captured.stdout + "\n\n" + self.captured.stderr

        return tmpl.format(
            outputs=outputs, results=display, code=self.result.info.raw_cell
        )

    @staticmethod
    def output_to_html(output: "RichOutput | str") -> str:
        if isinstance(output, str):
            return output

        if (res := output._repr_mime_("image/png")) is not None:
            return f'<img src="data:image/png;base64,{res[0]}" />'
        elif (res := output._repr_mime_("text/html")) is not None:
            return res[0]

        return output.data.get("text/plain", "")

    @staticmethod
    def output_to_content(output: RichOutput) -> Content:
        if (res := output._repr_mime_("image/png")) is not None:
            # res is data and metadata
            return ContentImageInline("image/png", res[0])
        elif (res := output._repr_mime_("text/html")) is not None:
            return ContentText("text/html", res[0])

        return ContentText(output.data.get("text/plain", ""))


def run_cell_capture_exceptions(shell: InteractiveShell, code: str) -> CellRun:
    from IPython.utils.capture import capture_output

    # Create a variable to store the exception if one occurs
    exception_info = None

    # Define a custom error handler function
    def custom_handler(shell, etype, evalue, tb, tb_offset=None):
        nonlocal exception_info
        exception_info = (etype, evalue, tb)
        # return shell.InteractiveTB.handler(etype, evalue, tb, tb_offset)

    # Store the original exception handler
    original_handler = shell.custom_exceptions

    try:
        # Set our custom handler
        shell.set_custom_exc((Exception,), custom_handler)

        # Run the cell with output capture
        with capture_output() as captured:
            result = shell.run_cell(code)

        return CellRun(captured, result, exception_info)
    finally:
        # Restore the original handler
        shell.custom_exceptions = original_handler

        # TODO: handle worst case CellRun return


def execute_code(code: str, as_tool: bool = True):
    """Execute python code in the current IPython shell."""

    from IPython.core.getipython import get_ipython
    from IPython.display import display, HTML

    shell = get_ipython()

    if shell is None:
        raise ValueError("Result of get_ipython() is None.")

    cell_run = run_cell_capture_exceptions(shell, code)
    # with capture_output() as cap:
    #    res = shell.run_cell(code, store_history=False)

    if as_tool:
        display(HTML(cell_run.to_html()))
        return cell_run.to_tool_result()

    return cell_run
