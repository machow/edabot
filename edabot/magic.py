from IPython.core.magic import Magics, magics_class, line_cell_magic
from IPython.core.magic_arguments import magic_arguments, argument, parse_argstring
from IPython.core.interactiveshell import InteractiveShell


@magics_class
class ChatMagics(Magics):
    shell: InteractiveShell

    @magic_arguments()
    @argument("-c", "--chat", help="Chat variable name", default="chat")
    @argument("prompt", help="Prompt to send to the chat", nargs="*")
    @line_cell_magic
    def chat(self, line, cell=None):
        "Magic that works both as %lcmagic and as %%lcmagic"
        args = parse_argstring(self.chat, line)

        chat = self.shell.user_ns.get(args.chat, None)

        if chat is None:
            raise ValueError(f"Chat variable '{args.chat}' not found in user namespace.")

        if cell is None:
            print(line)
            chat.chat(" ".join(args.prompt))
        else:
            print(line)
            print(cell)
            chat.chat(cell)

