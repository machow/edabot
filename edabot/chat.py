from __future__ import annotations

import os
from chatlas import ChatAnthropic
from ._chatlas import Chat2
from dotenv import find_dotenv, dotenv_values

from .execute import execute_code


def create_edabot() -> Chat2:
    chat = create_chat()
    chat.register_tool(execute_code)
    return chat


def create_chat() -> Chat2:
    env = dotenv_values(find_dotenv())
    api_key = env.get("ANTHROPIC_API_KEY", os.environ.get("ANTHROPIC_API_KEY"))
    if api_key is None:
        raise ValueError(
            "ANTHROPIC_API_KEY not found in environment variables or .env file."
        )
    chat = ChatAnthropic(
        model="claude-3-5-sonnet-20241022",
        system_prompt="You are a friendly but terse assistant.",
        api_key=api_key,
    )

    chat.__class__ = Chat2
    return chat
