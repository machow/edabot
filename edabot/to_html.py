from functools import singledispatch
from .execute import CellRun

@singledispatch
def to_html(obj) -> str:
    """Convert an object to HTML."""
    raise NotImplementedError(f"Cannot convert {type(obj)} to HTML.")

@to_html.register
def _(obj: CellRun) -> str:
    style = r"""
details.edabot-cell-run {
  /*background-color: #fbebfc;*/
  border: 2px solid #fbebfc;
  border-radius: 2px;
  & summary {
    cursor: pointer;
    padding: 10px;
    background-color: #fbebfc;
  }
}

.edabot-cell-run {
  pre {
    background-color: #f9f9f9;
    padding: 5px;
    max-height: 300px;
    overflow: scroll;
  }

  & .edabot-cell-run__section {
    padding: 5px;
  }
}
"""
    tmpl = """
<style>
{style}
</style>
<details class="edabot-cell-run">
    <summary>Execute Code</summary>
    <div class="edabot-cell-run__code edabot-cell-run__section">
      <h3>Code</h3>
      <pre>{code}</pre>
    </div>
    <div class="edabot-cell-run__outputs edabot-cell-run__section">
      <h3>Outputs</h3>
      <pre>{outputs}</pre>
    </div>
    <div class="edabot-cell-run__display edabot-cell-run__section">
      <h3>Display</h3>
      <div class="display-results">{results}</div>
    </div>
</details>
    """
    display = "\n\n".join(map(obj.output_to_html, obj.captured.outputs))

    outputs = obj.captured.stdout + "\n\n" + obj.captured.stderr

    return tmpl.format(
        style=style, outputs=outputs, results=display, code=obj.result.info.raw_cell
    )
