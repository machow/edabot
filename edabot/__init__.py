from ._chatlas import ctx_display
from .chat import create_chat, create_edabot
from .execute import execute_code
from .to_html import to_html

def load_ipython_extension(ipython):
    from .magic import ChatMagics
    ipython.register_magics(ChatMagics)