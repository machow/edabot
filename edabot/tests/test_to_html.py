from edabot import to_html, execute_code

def test_to_html_execute_code(gshell, snapshot):
    res = execute_code("print('hey')\n1", as_tool=False)
    html = to_html(res)
    assert snapshot == html
