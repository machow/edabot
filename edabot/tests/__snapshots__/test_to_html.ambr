# serializer version: 1
# name: test_to_html_execute_code
  '''
  
  <style>
  
  details.edabot-cell-run {
    /*background-color: #fbebfc;*/
    border: 2px solid #fbebfc;
    border-radius: 2px;
    & summary {
      cursor: pointer;
      padding: 10px;
      background-color: #fbebfc;
    }
  }
  
  .edabot-cell-run {
    pre {
      background-color: #f9f9f9;
      padding: 5px;
      max-height: 300px;
      overflow: scroll;
    }
  
    & .edabot-cell-run__section {
      padding: 5px;
    }
  }
  
  </style>
  <details class="edabot-cell-run">
      <summary>Execute Code</summary>
      <div class="edabot-cell-run__code edabot-cell-run__section">
        <h3>Code</h3>
        <pre>print('hey')
  1</pre>
      </div>
      <div class="edabot-cell-run__outputs edabot-cell-run__section">
        <h3>Outputs</h3>
        <pre>hey
  Out[1]: 1
  
  
  </pre>
      </div>
      <div class="edabot-cell-run__display edabot-cell-run__section">
        <h3>Display</h3>
        <div class="display-results"></div>
      </div>
  </details>
      
  '''
# ---
